// mkdist/dist/chunks/make.mjs
import globby from 'globby';
import { resolve, extname, basename, join, dirname } from 'pathe';
import fse from 'fs-extra';
import { transform } from 'esbuild';
import jiti from 'jiti';

const DECLARATION_RE = /\.d\.[cm]?ts$/;
const CM_LETTER_RE = /(?<=\.)(c|m)(?=[tj]s$)/;
const jsLoader = async (input, { options }) => {
  if (![".ts", ".js", ".cjs", ".mjs"].includes(input.extension) || input.path.match(DECLARATION_RE)) {
    return;
  }
  const output = [];
  let contents = await input.getContents();
  if (options.declaration && !input.srcPath?.match(DECLARATION_RE)) {
    const cm = input.srcPath?.match(CM_LETTER_RE)?.[0] || "";
    const extension = `.d.${cm}ts`;
    output.push({
      contents,
      srcPath: input.srcPath,
      path: input.path,
      extension,
      declaration: true
    });
  }
  if (input.extension === ".ts") {
    contents = await transform(contents, { loader: "ts" }).then((r) => r.code);
  }
  const isCjs = options.format === "cjs";
  if (isCjs) {
    contents = jiti().transform({ source: contents, retainLines: false }).replace(/^exports.default = /mg, "module.exports = ");
  }
  output.push({
    contents,
    path: input.path,
    extension: options.ext ? `.${options.ext}` : isCjs ? ".js" : ".mjs"
  });
  return output;
};

const vueLoader = async (input, { loadFile }) => {
  if (input.extension !== ".vue") {
    return;
  }
  const contents = await input.getContents();
  const [scriptBlock, attrs = "", _lastAttr, script] = contents.match(/<script((\s[^>\s]*)*)>([\S\s.]*?)<\/script>/) || [];
  if (!scriptBlock || !script) {
    return;
  }
  if (attrs.split(/\s+/g).includes("setup")) {
    return;
  }
  const [, lang = "js"] = attrs.match(/lang="([a-z]*)"/) || [];
  const extension = "." + lang;
  const files = await loadFile({
    getContents: () => script,
    path: `${input.path}${extension}`,
    srcPath: `${input.srcPath}${extension}`,
    extension
  }) || [];
  const scriptFile = files.find((f) => [".js", ".mjs"].includes(f.extension));
  if (!scriptFile) {
    return;
  }
  const newAttrs = attrs.replace(new RegExp(`\\s?lang="${lang}"`), "");
  return [
    {
      path: input.path,
      contents: contents.replace(scriptBlock, `<script${newAttrs}>
${scriptFile.contents}<\/script>`)
    },
    ...files.filter((f) => f !== scriptFile)
  ];
};

const defaultLoaders = [vueLoader, jsLoader];
function createLoader(loaderOptions = {}) {
  const loaders = loaderOptions.loaders || defaultLoaders;
  const loadFile = async function(input) {
    const context = {
      loadFile,
      options: loaderOptions
    };
    for (const loader of loaders) {
      const outputs = await loader(input, context);
      if (outputs?.length) {
        return outputs;
      }
    }
    return [
      {
        path: input.path,
        srcPath: input.srcPath,
        raw: true
      }
    ];
  };
  return {
    loadFile
  };
}

async function getDeclarations(vfs) {
  const ts = await import('typescript').then((r) => r.default || r);
  const inputFiles = [...vfs.keys()];
  const compilerOptions = {
    allowJs: true,
    declaration: true,
    incremental: true,
    skipLibCheck: true,
    emitDeclarationOnly: true
  };
  const tsHost = ts.createCompilerHost(compilerOptions);
  tsHost.writeFile = (fileName, declaration) => {
    vfs.set(fileName, declaration);
  };
  const _readFile = tsHost.readFile;
  tsHost.readFile = (filename) => {
    if (vfs.has(filename)) {
      return vfs.get(filename);
    }
    return _readFile(filename);
  };
  const program = ts.createProgram(inputFiles, compilerOptions, tsHost);
  await program.emit();
  const output = {};
  for (const filename of inputFiles) {
    const dtsFilename = filename.replace(/\.(m|c)?(ts|js)$/, ".d.$1ts");
    output[filename] = vfs.get(dtsFilename) || "";
  }
  return output;
}

async function mkdist(options = {}) {
  options.rootDir = resolve(process.cwd(), options.rootDir || ".");
  options.srcDir = resolve(options.rootDir, options.srcDir || "src");
  options.distDir = resolve(options.rootDir, options.distDir || "dist");
  if (options.cleanDist !== false) {
    await fse.unlink(options.distDir).catch(() => {
    });
    await fse.emptyDir(options.distDir);
    await fse.mkdirp(options.distDir);
  }
  const filePaths = await globby(options.pattern || "**", { absolute: false, cwd: options.srcDir });
  const files = filePaths.map((path) => {
    const srcPath = resolve(options.srcDir, path);
    return {
      path,
      srcPath,
      extension: extname(path),
      getContents: () => fse.readFile(srcPath, { encoding: "utf8" })
    };
  });
  const { loadFile } = createLoader({
    format: options.format,
    ext: options.ext,
    declaration: options.declaration
  });
  const outputs = [];
  for (const file of files) {
    outputs.push(...await loadFile(file) || []);
  }
  for (const output of outputs.filter((o) => o.extension)) {
    const renamed = basename(output.path, extname(output.path)) + output.extension;
    output.path = join(dirname(output.path), renamed);
    if (outputs.find((o) => o !== output && o.path === output.path)) {
      output.skip = true;
    }
  }
  const dtsOutputs = outputs.filter((o) => o.declaration && !o.skip);
  if (dtsOutputs.length) {
    const declarations = await getDeclarations(new Map(dtsOutputs.map((o) => [o.srcPath, o.contents || ""])));
    for (const output of dtsOutputs) {
      output.contents = declarations[output.srcPath] || "";
    }
  }
  const outPaths = new Set(outputs.map((o) => o.path));
  const resolveId = (from, id = "", resolveExts) => {
    if (!id.startsWith(".")) {
      return id;
    }
    for (const ext of resolveExts) {
      if (outPaths.has(join(dirname(from), id + ext))) {
        return id + ext;
      }
    }
    return id;
  };
  const esmResolveExts = ["", "/index.mjs", "/index.js", ".mjs", ".ts"];
  for (const output of outputs.filter((o) => o.extension === ".mjs")) {
    output.contents = output.contents.replace(/(import|export)(.* from ['"])(.*)(['"])/g, (_, type, head, id, tail) => type + head + resolveId(output.path, id, esmResolveExts) + tail);
  }
  const cjsResolveExts = ["", "/index.cjs", ".cjs"];
  for (const output of outputs.filter((o) => o.extension === ".cjs")) {
    output.contents = output.contents.replace(/require\((['"])(.*)(['"])\)/g, (_, head, id, tail) => "require(" + head + resolveId(output.path, id, cjsResolveExts) + tail + ")");
  }
  const writtenFiles = [];
  await Promise.all(outputs.filter((o) => !o.skip).map(async (output) => {
    const outFile = join(options.distDir, output.path);
    await fse.mkdirp(dirname(outFile));
    if (output.raw) {
      await fse.copyFile(output.srcPath, outFile);
    } else {
      await fse.writeFile(outFile, output.contents, "utf8");
    }
    writtenFiles.push(outFile);
  }));
  return {
    writtenFiles
  };
}

export { mkdist as m };
